core/src/include/ModelArray.hpp

```diff
@@ -330,7 +330,12 @@ class ModelArray {
      * @param type The type of array the dimensions are to be specified for.
      * @param dim The per-dimension size to be set.
      */
+#ifdef USE_MPI
+    static void setDimensions(Type, const MultiDim&, const MultiDim&);
```

Comment:
I must admit that I don't like the idea of having a different function signature serial and MPI versions of the code. I wonder if a better solution might be changing the result of `ModelArray::dimensions` from a plain `MultiDim` to a vector of not scalars but local-global pairs in the MPI case.